Add Instana exporter implementation

hickeyma opened the pull request:
Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Martin Hickey [martin.hickey@ie.ibm.com](mailto:martin.hickey@ie.ibm.com)
Co-authored-by: Cedric Ziel [cedric@cedric-ziel.com](mailto:cedric@cedric-ziel.com)
Co-authored-by: Willian Carvalho [willianc@pm.me](mailto:willianc@pm.me)

**Description:** <Describe what has changed.>

This is the second of 3 PRs which adds the  new component, Instana exporter. This PR
adds the exporter implementation.

**Link to tracking Issue:** <Issue number if applicable>

Partial https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13395

**Testing:** 

Unit Tests

hickeyma committed:
Add Instana exporter implementation

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add unit tests

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add changelog

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,164 @@
+func (e *instanaExporter) pushConvertedTraces(ctx context.Context, td ptrace.Traces) error {
+	e.settings.Logger.Info("TracesExporter", zap.Int("#spans", td.SpanCount()))
+	if !e.settings.Logger.Core().Enabled(zapcore.DebugLevel) {
+		return nil
```
Is this intended to be here?

hickeyma replied:
No, good catch. Removed.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,164 @@
+
+	buf, err := e.tracesMarshaler.MarshalTraces(td)
+	if err != nil {
+		return err
```
This sounds like a permanent error. Use consumererror.NewPermanent to wrap this: https://github.com/open-telemetry/opentelemetry-collector/blob/e092fc728b724f862d37294dff03a602996f1dd2/consumer/consumererror/permanent.go#L27-L29

This way, this won't be retried.

hickeyma replied:
Sure. Removing as unused for now.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		return err
+	}
+	e.settings.Logger.Debug(string(buf))
```
This does not belong here.

hickeyma replied:
Throwback from debugging, thanks. Removed.

jpkrohling reviewed (commented):
Before I continue reviewing this, could you please go over the code and remove the debug statements that will not belong to the final version?

hickeyma committed:
Add more unit tests

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+func (e *instanaExporter) pushConvertedTraces(ctx context.Context, td ptrace.Traces) error {
+	e.settings.Logger.Info("TracesExporter", zap.Int("#spans", td.SpanCount()))
```
Hmm, good idea. Going to remove it for now and can be looked again when metrics are introduced.

hickeyma committed:
Update fater review

Review comments:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#pullrequestreview-1085527736

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thank you for the review and feedback @jpkrohling. I have updated your comments and it is ready for review again.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,144 @@
+
+		resource := resSpan.Resource()
+
+		hostIDAttr, ex := resource.Attributes().Get(backend.AttributeInstanaHostID)
```
It took me a while to realize that "ex" probably means "exists". Perhaps use `ok` instead? Or even `exists`?

hickeyma replied:
Done

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,144 @@
+
+	bundle := model.Bundle{Spans: spans}
+
+	if len(bundle.Spans) == 0 {
```
Would this commonly happen? I would expect a warning message to exist somewhere if spans are being received but not exported.

hickeyma replied:
Probably not but is it better incase such a scenario arises we are unaware of now?

hickeyma replied:
What do you think?

jpkrohling replied:
I guess there are two situations: your component received an empty batch, in which case, it's OK to have an empty outcome. I'm not sure we _should_ catch this in the component, though. 

Then, there's the situation where you get a non-empty batch, but you filter out spans and end up with an empty batch. In that case, it would be good to have a metric stating how many spans were received and how many were filtered out. Users can then use those numbers to debug and understand what's going on.

It doesn't look like you are filtering anything, though, so I can't see whether this case will ever happen. If you are indeed not filtering anything, just get rid of this condition.

jpkrohling replied:
If you prefer to deal with this in a subsequent PR, that's fine by me.

hickeyma replied:
Ok, will think about it and deal with it in another PR.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,144 @@
+	headers := map[string]string{
+		backend.HeaderKey:  e.config.AgentKey,
+		backend.HeaderHost: hostID,
+		backend.HeaderTime: "0",
```
This deserves an explanation in the form of a code comment.

hickeyma replied:
Investigating this. I will add a comment shortly

hickeyma replied:
Comment added.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,144 @@
+func newInstanaExporter(cfg config.Exporter, set component.ExporterCreateSettings) (*instanaExporter, error) {
+	iCfg := cfg.(*Config)
+
+	if iCfg.Endpoint != "" {
```
Don't you have validation for this already as part of the config object?

hickeyma replied:
Nice catch. Fixed now and removed from there.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,144 @@
+
+	resp, err := e.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to make a HTTP request: %w", err)
```
You could optimize this a bit: if this is a 4xx error type, it's a permanent error; if it's a 5xx error type, it's a regular error.

jpkrohling commented on exporter/instanaexporter/exporter.go:
```diff
@@ -0,0 +1,144 @@
+		return fmt.Errorf("failed to make a HTTP request: %w", err)
+	}
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
```
I read this as:

```
if resp... {
   return nil
} else {
   return nil 
}
```

Perhaps just get rid of this? Or return an error on the else clause?

hickeyma replied:
Removed

jpkrohling commented on exporter/instanaexporter/exporter_test.go:
```diff
@@ -0,0 +1,104 @@
+	got := make(chan string, 1)
+	traceServer := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		assert.Equal(t, "key11", req.Header.Get("x-instana-key"))
+		got <- req.Header.Get("Content-Type")
```
Why do you have this here?

hickeyma replied:
Spurious text unfortunately. Removed.

jpkrohling commented on exporter/instanaexporter/internal/converter/all_converter.go:
```diff
@@ -0,0 +1,69 @@
+
+	for i := 0; i < len(c.converters); i++ {
+		if !c.converters[i].AcceptsSpans(attributes, spanSlice) {
+			c.logger.Warn(fmt.Sprintf("Converter %s didnt Accept", c.converters[i].Name()))
```
```suggestion
			c.logger.Warn(fmt.Sprintf("Converter %q didn't accept spans", c.converters[i].Name()))
```

hickeyma replied:
Updated

jpkrohling commented on exporter/instanaexporter/internal/converter/model/util.go:
```diff
@@ -0,0 +1,73 @@
+	return hex.EncodeToString(spanBytes)
+}
+
+func oTelKindToInstanaKind(otelKind ptrace.SpanKind) (string, bool) {
```
```suggestion
func otelKindToInstanaKind(otelKind ptrace.SpanKind) (string, bool) {
```

hickeyma replied:
Updated

jpkrohling commented on exporter/instanaexporter/internal/otlptext/databuffer.go:
```diff
@@ -0,0 +1,152 @@
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+type dataBuffer struct {
```
This code seemed familiar, and I found it here:

https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/loggingexporter/internal/otlptext/databuffer.go

I would either suggest splitting the original otlptext package into a public utility package or at least leaving a comment pointing people to the original resource so that any fixes (especially security) could be ported to this version.

hickeyma replied:
Thanks for catching this. As this code isn't originally mine, I hadn't realised that it was copied. 

For the moment, I have added a comment to the top of the file specifying its a copy and to keep uptodate. In medium to long term, we need to either refactor the code to avoid the duplication oro work to expose the collector code. Whichever makes most sense.

hickeyma committed:
Update after review

Review comments:

- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957596160
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957603426
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957603426
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957609409
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957610214
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957612370
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957613538
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957626028
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13620#discussion_r957629599

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks for the review @jpkrohling. Updated and ready for review again.
